rllib/agents/dqn/dqn.py

```diff
@@ -271,6 +296,117 @@ def update_prio(item):
 
         return StandardMetricsReporting(train_op, workers, config)
 
+    @ExperimentalAPI
+    def training_iteration(self) -> ResultDict:
+        """DQN training iteration function.
+
+        This replicates DQN's execution_plan behaviour, e.g.:
+        - For every time we (1) sample (MultiAgentBatch) from workers...
+            - (2) Concatenate freshly collected samples.
+            - (3) Store new samples in replay buffer.
+            - (4) Sample training batch (MultiAgentBatch) from replay buffer.
+            - (5) Learn on training batch.
+            - (6) Update target network every target_network_update_freq steps.
+        - (7) Return all collected metrics for the iteration.
+
+        Returns:
+            The results dict from executing the training iteration.
+        """
+        batch_size = self.config["train_batch_size"]
+        local_worker = self.workers.local_worker()
+
+        sample_batches = []
+        train_results = {}
+
+        # We alternate between storing new samples and sampling and training
+        store_weight, sample_and_train_weight = calculate_rr_weights(self.config)
+
+        for _ in range(store_weight):
+            # (1) Sample (MultiAgentBatch) from workers
+            new_sample_batches = synchronous_parallel_sample(self.workers)
+            sample_batches.extend(new_sample_batches)
+
+            # Update counters
+            self._counters[NUM_ENV_STEPS_SAMPLED] += sum(
+                len(s) for s in new_sample_batches
+            )
+            self._counters[NUM_AGENT_STEPS_SAMPLED] += sum(
+                len(s) if isinstance(s, SampleBatch) else s.agent_steps()
+                for s in new_sample_batches
+            )
+
+            # (2) Concatenate freshly collected samples
+            concatenated_samples = SampleBatch.concat_samples(sample_batches)
```

Comment:
Shouldn't this be: `new_sample_batches` here? Otherwise, we would add previous ones (from the previous for-iteration) more than once, no?

Reply:
Wow that's a big one! Thanks!